hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/FileSystemBasedLockProvider.java

```diff
@@ -139,7 +152,32 @@ private boolean checkIfExpired() {
 
   private void acquireLock() {
     try {
-      fs.create(this.lockFile, false).close();
+      if (!fs.exists(this.lockFile)) {
+        FSDataOutputStream fos = fs.create(this.lockFile, false);
+        initLockInfo();
+        fos.writeBytes(lockInfo.toString());
```

Comment:
May I know if you guys tested against both hdfs and S3.

Atleast in S3, not sure if this works across all race conditions.

what if two processes exactly checks fs.exists() concurrently and both detects that file does not exist and goes ahead and creates the file. both creation could succeed. So, how do we ensure only one of them succeed.

Should we compare the file contents after creating. so that the current writer which wrote the file will exactly know that it owns the lock if file contents match.

Reply:
>> May I know if you guys tested against both hdfs and S3.

I did not test on S3, just used this feature(on HDFS) only several times. @fengjian428 Can you give it a try?

If it occurs race conditions on S3, we should find a way to fix it.

Reply:
@nsivabalan  I'm not sure about that since no S3 environment on my side. I supposed there should be only one process to succeed, if not, we can write InstantTime into file content and check the file content after obtaining the lock as you said. I can add this logic, it should have no side effect

> May I know if you guys tested against both hdfs and S3.
> 
> Atleast in S3, not sure if this works across all race conditions.
> 
> what if two processes exactly checks fs.exists() concurrently and both detects that file does not exist and goes ahead and creates the file. both creation could succeed. So, how do we ensure only one of them succeed.
> 
> Should we compare the file contents after creating. so that the current writer which wrote the file will exactly know that it owns the lock if file contents match.

Reply:
sorry, I think adding instant time into file content cannot guarantee just one process obtain the lock either.  the other process still can overwrite the file content after the first process already validated the file content. file-based lock  only works for DFS with atomic create/delete operation

Reply:
https://github.com/apache/hudi/pull/7614 check if dfs support atomic creation when using the filesystem base lock provider

Reply:
+1